src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1493,19 +1493,24 @@
 
 ;; Wrap the fact-type so that Clojure equality and hashcode semantics are used
 ;; even though this is placed in a Java map.
-(deftype AlphaRootsWrapper [roots roots-hash]
+(deftype AlphaRootsWrapper [fact-type fact-type-hash roots]
   Object
   (equals [this other]
-    ;; Since we memoize the function that retrieves these on the fact type, and since the alpha roots for
-    ;; any given type primary type T (excluding descendants and ancestors) are a disjoint
-    ;; set with respect to the alpha roots for any other primary type T2, we know that we will
-    ;; only have one AlphaRootsWrapper object for any given set of alpha roots per get-alphas-fn instance.
-    ;; Therefore we can just use reference equality in the interest of maximizing performance.
-    (identical? this other))
+    (let [other ^AlphaRootsWrapper other]
+      (cond
+        
+        (identical? fact-type (.fact-type other)) 
+        true
+        
+        (not (== fact-type-hash (.fact-type-hash other)))
```

Comment:
This would be faster if you give a primitive type hint `^int` on the `deftype` signature for `fact-type-hash`.

I mentioned this in a previous version.
```clj
(deftype AlphaRootsWrapper [fact-type ^int fact-type-hash roots] <etc>)
```